REVIEWER VERSION

Stampyswamp opened the pull request:
Added All classes + finished each. (Had to delete old repo. accidentally pushed the changes to master branch)
Each class has all the advanced Functionality Implemented.

Stampyswamp committed:
FINAL.

paulfrancis556 commented on CMP1903_A1_2324/Game.cs:
```diff
@@ -1,21 +1,41 @@
+            /*Rolls Dice
+            Simple loop while roll the dice x amount of times, adding each roll to a list. */
+            while (i <= amountToRoll)
+            {
```
This block of code is a bit unreadable at first glance. Id recommend adding comments and some extra line breaks to make the code more readable

Stampyswamp replied:
**DONE**

paulfrancis556 commented on CMP1903_A1_2324/Die.cs:
```diff
@@ -1,23 +1,18 @@
+        /// Produces a random roll 1-6. if (parameter is changed && and num > 6 is rolled) {debug class will flag this.} (not real code)
+        /// </summary>
+        /// <returns> Returns roll which is a random number 1-6. </returns> 
+        public int throwDice()
```
The specification provided required that this method is called Roll()

paulfrancis556 commented on CMP1903_A1_2324/Game.cs:
```diff
@@ -1,21 +1,41 @@
+                Console.WriteLine($"Dice " + i + ": " + roll);
+                debug.Assert(roll);
+                list.Add(roll);
+                System.Threading.Thread.Sleep(1); ///Sleep allows program to Create new Object.
```
Using this is non-sensical.
Random() relies on the system time to create a seed to create a random number from. 

Your script creates a new Random() object every time which is where the real problem is.

If you change where/how you create your Random() object to static in die.cs, it will use a single random object for all instances.

By using Sleep() you run the risk of your code becoming uncessarily very slow as the code functionality is expanded - for example - if you asked your code to roll 1000 dice - this would take 1 second, as your code is sleeping 1 millisecond every time.

paulfrancis556 replied:
For furthur clarity, take a look at https://stackoverflow.com/questions/1654887/random-next-returns-always-the-same-values

paulfrancis556 replied:
Final Note: If you're ever intentionally slowing your software down, its generally a good sign that you can refactor the code so you dont have to do that - our mission in life is to make software quick! not slow :) (Users hate slow software...)

paulfrancis556 commented on CMP1903_A1_2324/Die.cs:
```diff
@@ -1,23 +1,18 @@
-        //Method
-
-
+        Random rnd = new Random();
```
Change this to "private static Random rnd = new Random();" - See my comment below where you have used "                System.Threading.Thread.Sleep(1);" in game.cs as to why.

paulfrancis556 commented on CMP1903_A1_2324/Die.cs:
```diff
@@ -1,23 +1,18 @@
-         * and one method that rolls the die, returns and integer and takes no parameters.
-         */
-
-        //Property
```
The specificatiom says 'The Die class should contain one property to hold the die current value'

Using get/set would create a property in the correct way.

Please see https://www.w3schools.com/cs/cs_properties.php for further clarification

Stampyswamp replied:
**DONE**

paulfrancis556 commented on CMP1903_A1_2324/Game.cs:
```diff
@@ -1,21 +1,41 @@
+            int amountToRoll = 3;
+            /*Rolls Dice
+            Simple loop while roll the dice x amount of times, adding each roll to a list. */
+            while (i <= amountToRoll)
```
Using a for rather than a while will shorten the lines of code slightly, and make the code a bit more readable.

**for (int i = 0; i < 3; i++){
}**

 will give you the same result as

**int i = 1;
int amountToRoll = 3;
while (i <= amountToRoll){
i++;
}**

paulfrancis556 commented on CMP1903_A1_2324/Game.cs:
```diff
@@ -1,21 +1,41 @@
+            Each diceroll in the list will be added to the sum.*/
+            Testing sumCheck = new Testing();
+            int sum = 0;
+            foreach (var item in list) { sum += item; }
```
You can use list.Sum() to sum all the values in a list

paulfrancis556 commented on CMP1903_A1_2324/Testing.cs:
```diff
@@ -1,20 +1,44 @@
+        /// </summary>
+        /// <param name="roll"></param>
+        /// <returns> Writes in the console if the role is legal, else returns true/false incase you want to do something with this in the future. as no errors are raised. </returns>
+        public bool Assert(int roll)
```
The specification asked us to use debug.assert()

Youve created your own asset class. 

c# has a debug.assert() method built in as part of the System.Diagnostics namespace. 

See https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.debug.assert?view=net-8.0 for more details

Stampyswamp committed:
Update Die.cs

static object added

Stampyswamp committed:
Update Program.cs

removed time.sleep()

Stampyswamp committed:
Update Game.cs

Stampyswamp committed:
Update Die.cs

Roll

Stampyswamp committed:
Update Game.cs

Swapped Function name to match

Stampyswamp committed:
Update Die.cs

Changed roll to a private encapsulated variable.

Stampyswamp committed:
Update Game.cs

Changed while to for loop, removed unnecessary sleep + moved objects outside of for loop and made static.

Stampyswamp committed:
Update Testing.cs

changed the method to use debug.assert

Stampyswamp committed:
Update Game.cs

fixed debug.assert

Stampyswamp committed:
Update Game.cs

changed for loop so the dice printed rolls correctly.

Stampyswamp committed:
Update Game.cs

Tested code and fixed problems during runtime.

Stampyswamp committed:
Update Program.cs

Double Check.

Stampyswamp committed:
Update Game.cs

Fixed threading sleep. AGAIN.

Stampyswamp committed:
Update Testing.cs

fixed minor comment bugs.

ML-04 commented on CMP1903_A1_2324/Die.cs:
**Positive**
+ The comments provided give a good explanation of how things work.
+ The code is clear and concise.

**Negative**
- Display use of the Encapsulation. 
You can see the documentation here: https://www.w3schools.com/cs/cs_properties.php

ML-04 commented on CMP1903_A1_2324/Game.cs:
**Positive**
+ There's lots of comments that provide information of what each thing does.

ML-04 commented on CMP1903_A1_2324/Program.cs:
**Positive**
+ Not only do the comments provide a good explanation but there are many of them to provide what everything does.
+ Good use of coloured text for printing out information in the console.

ML-04 commented on CMP1903_A1_2324/Testing.cs:
**Positive**
+ The code is clear and concise.

Stampyswamp committed:
Update Die.cs
